src/metabase/query_processor/middleware/expand_macros.clj

```diff
@@ -34,11 +34,23 @@
         (throw (IllegalArgumentException. (tru "Segment {0} does not exist, or is invalid." segment-id))))))
 
 (s/defn ^:private expand-segments :- mbql.s/Query
-  [{inner-query :query, :as outer-query} :- mbql.s/Query]
-  (if-let [segments (mbql.u/match inner-query :segment)]
-    (replace-segment-clauses outer-query (segment-clauses->id->definition segments))
-    outer-query))
-
+  "Recursively expand segments in the `query`."
+  [query :- mbql.s/Query]
+  (loop [{inner-query :query :as outer-query} query
+         depth 0]
+    (if-let [segments (mbql.u/match inner-query [:segment (_ :guard (complement mbql.u/ga-id?))])]
+      (let [segment-id->definition (segment-clauses->id->definition segments)
+            expanded-query (replace-segment-clauses outer-query segment-id->definition)]
+        ;; Following line is in place to avoid infinite recursion caused by mutually recursive
+        ;; segment definitions or other unforseen circumstances. Number 41 is arbitrary.
+        (if (or (= expanded-query outer-query) (= depth 41))
+          (throw (ex-info (tru "Segment expansion failed. Check mutually recursive segment definitions.")
+                          {:original-query query
+                           :expanded-query expanded-query
+                           :segment-id->definition segment-id->definition
+                           :depth depth}))
```

Comment:
Please add a `:type` with a `qp.error-type/` in here, we do use it for some stuff like deciding whether a query should be retried or not.

Reply:
Using `qp.error-type/invalid-query` now.